### PR TITLE
support pickle/unpickle for typed dicts

### LINKED ISSUE #377: `TypedDict` is not serializable
traceback from the original issue
```
    return base64.b64encode(dill.dumps(obj)).decode('utf-8')
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 263, in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 235, in dump
    Pickler(file, protocol, **_kwds).dump(obj)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 394, in dump
    StockPickler.dump(self, obj)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 487, in dump
    self.save(obj)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 717, in save_reduce
    save(state)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 1186, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 971, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 997, in _batch_setitems
    save(v)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 931, in save_list
    self._batch_appends(obj)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 955, in _batch_appends
    save(x)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 717, in save_reduce
    save(state)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 1186, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 971, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 997, in _batch_setitems
    save(v)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 717, in save_reduce
    save(state)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 1186, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 971, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 997, in _batch_setitems
    save(v)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/dill/_dill.py", line 388, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/opt/cray/pe/python/3.9.12.1/lib/python3.9/pickle.py", line 578, in save
    rv = reduce(self.proto)
  File "/autofs/nccs-svm1_home1/sharplo/venv_rct/lib/python3.9/site-packages/radical/utils/typeddict.py", line 270, in __getattr__
    else            : return data[k]
KeyError: '__getstate__'
```

### PR DESCRIPTION
**Pull request by andre-merzky**
This   fixes #377

### CODE REVIEW

File: src/radical/utils/typeddict.py [hunk 1]
```diff
@@ -262,6 +261,14 @@ def popitem(self):
     #
     def __getattr__(self, k):
 
+        if k == '_data':
+            if '_data' not in self.__dict__:
+                self.__dict__['_data'] = dict()
+            return self.__dict__['_data']
```
Comment (mtitov) [inline src/radical/utils/typeddict.py:267]: small proposal to have it (L265-L267) as `return self.__dict__.setdefault('_data', {})`

**Review by mtitov — APPROVED** [review]
LGTM! have just small suggestion

File: src/radical/utils/typeddict.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/utils/typeddict.py:267]: Semantically the same, but IMHO less readable: `setdefault` is rarely used, and I dislike that a `set` method returns things it is not setting - bah!   So I'd rather leave as is.   Does that work for you?

Comment (mtitov) [reply src/radical/utils/typeddict.py:267]: yeah, good to go!

**Pull request merged**